### PR TITLE
Add PPCR metric calculation

### PR DESCRIPTION
**Pull request by zurk**
Closes https://github.com/src-d/style-analyzer/issues/484

### CODE REVIEW

File: lookout/style/format/benchmarks/top_repos_quality.py [hunk 1]
```diff
@@ -308,18 +309,18 @@ def main(args):
...
                 n_rules, avg_len = _get_model_summary(report.model)
-                agg.append((precision, recall, f1, support, n_rules, avg_len))
-                table.append((get_repo_name(repo), precision, recall, f1, support, n_rules,
-                              avg_len))
+                row = metrics + (n_rules, avg_len)
+                agg.append(row)
+                table.append((get_repo_name(repo),) + row)
             # weighted average
-            table.append(("Weighted average", *("%.2f" % calc_weighted_avg(agg, i)
+            table.append(("Weighted average", *("%.6f" % calc_weighted_avg(agg, i)
```
Comment (vmarkovtsev) [inline lookout/style/format/benchmarks/top_repos_quality.py]: 6 digits are too much. 3 are enough, we are not on Kaggle.

File: lookout/style/format/benchmarks/general_report.py [hunk 2]
```diff
@@ -72,34 +44,43 @@ def _load_jinja2_template(report_template_filename: str) -> jinja2.Template:
...
-        if gt != pred:
-            file_mispred.append(vn.path)
+    for vnode in vnodes:
+        if hasattr(vnode, "y_old") and vnode.y != vnode.y_old:
+            file_mispred.append(vnode.path)
     file_stat = Counter(file_mispred)
     to_show = file_stat.most_common()
     if max_files > 0:
         to_show = to_show[:max_files]
+    json_report = dict(cl_report=c_sorted,
```
Comment (vmarkovtsev) [inline lookout/style/format/benchmarks/general_report.py]: This is template overfitting. Feed in `ppcr` and write everything else in the template

Comment (zurk) [reply lookout/style/format/benchmarks/general_report.py]: outdated. I change it a little bit.

File: lookout/style/format/benchmarks/general_report.py [hunk 3]
```diff
@@ -72,34 +44,47 @@ def _load_jinja2_template(report_template_filename: str) -> jinja2.Template:
...
     # confusion matrix
-    mat = confusion_matrix(y, y_pred)
+    mat = confusion_matrix(y, y_pred_pure)
     # sort files by mispredictions
     file_mispred = []
-    for gt, pred, vn in zip(y, y_pred, vnodes_y):
-        if gt != pred:
-            file_mispred.append(vn.path)
+    for vnode in vnodes:
+        if hasattr(vnode, "y_old") and vnode.y != vnode.y_old:
```
Comment (vmarkovtsev) [inline lookout/style/format/benchmarks/general_report.py]:
```python
if vnode.y != getattr(vnode, "y_old", vnode.y)
```

File: lookout/style/format/benchmarks/general_report.py [hunk 4]
```diff
@@ -72,34 +44,47 @@ def _load_jinja2_template(report_template_filename: str) -> jinja2.Template:
...
         to_show = to_show[:max_files]
+    json_report = dict(cl_report=c_sorted,
+                       cl_report_full=c_report_full,
+                       ppcr=ppcr)
 
     template = _load_jinja2_template("quality_report.md.jinja2")
     # TODO(vmarkovtsev): move all the logic inside the template
-    res = template.render(language=language, ptr=ptr, cl_report=c_sorted, conf_mat=mat,
-                          target_names=target_names, files=to_show)
+    res = template.render(language=language, ptr=ptr, conf_mat=mat, target_names=target_names,
```
Comment (vmarkovtsev) [inline lookout/style/format/benchmarks/general_report.py]: No, this is no good.

1. No `json_report`. Write it inside.
2. No `**kwargs`, you should be explicit in this case.

File: lookout/style/format/templates/quality_report.md.jinja2 [hunk 5]
```diff
@@ -29,6 +37,6 @@
 <details>
     <summary>Machine-readable report</summary>
 ```json
-{{ cl_report | tojson }}
+{{ dict(cl_report=cl_report, cl_report_full=cl_report_full, ppcr=ppcr) | tojson }}
```
Comment (vmarkovtsev) [inline lookout/style/format/templates/quality_report.md.jinja2]: Come on :)

```suggestion
{
  "cl_report": {{ cl_report | tojson }},
  "cl_report_full": {{ cl_report_full | tojson }},
  "ppcr": {{ ppcr | tojson }}
}
```

Comment (zurk) [reply lookout/style/format/templates/quality_report.md.jinja2]: it is more reliable and easy to extend. I prefer to keep as is.

Comment (vmarkovtsev) [reply lookout/style/format/templates/quality_report.md.jinja2]: It is not more reliable and the same complexity of extension. However, it goes against idiomatic templates and requires an additional filter - `dict` - which is not included for a reason. Please change.

**Pull request merged**